apisix/plugins/traffic-split.lua

```diff
@@ -309,7 +314,8 @@ function _M.access(conf, ctx)
         return
     end
 
-    local rr_up, err = lrucache(weighted_upstreams, nil, new_rr_obj, weighted_upstreams)
+    local rr_up, err = lrucache(weighted_upstreams, nil, new_rr_obj, weighted_upstreams,
```

Comment:
Need to use `ctx.matched_route.value.upstream_id` as the extra key.

Reply:
you mean
```lua
local route_upstream_id = ctx.matched_route.value.upstream_id
local rr_up, err = lrucache(weighted_upstreams, nil, new_rr_obj, weighted_upstreams, route_upstream_id)
```
?

Reply:
update

Reply:
`lrucache` accepts an extra key as the second argument.

Reply:
Need to solve it.

Reply:
@tzssangglass 
Personally, we need to add `route_upstream_id` to the key as `route_upstream_id` and `weighted_upstreams` may change differently. As using `route_upstream_id` in the second argument will hit the bug I mention, maybe we can switch to use https://github.com/apache/apisix/blob/c530135c5a179c99637d52d99e0e8a45ce4cc8b8/apisix/plugins/limit-req.lua#L69?

Reply:
I tried it and it still doesn't work, because this plugin still need to control the weight by modifying the `upstream_id` in the route,  even if `core.lrucache.plugin_ctx` is used, since `route_upstream_id` is in the key and `route_upstream_id` will be overwritten by the `plugin_upstream_id`, it will result in two keys and inaccurate data obtained.

use `weighted_upstreams` as key because `weighted_upstreams` do not change.

Reply:
Try?
```diff
diff --git apisix/plugins/traffic-split.lua apisix/plugins/traffic-split.lua
index eb7a2028..966275f9 100644
--- apisix/plugins/traffic-split.lua
+++ apisix/plugins/traffic-split.lua
@@ -315,8 +315,8 @@ function _M.access(conf, ctx)
     end

     local route_upstream_id = ctx.matched_route.value.upstream_id
-    local rr_up, err = lrucache(weighted_upstreams, nil, new_rr_obj, weighted_upstreams,
-                                route_upstream_id)
+    local rr_up, err = core.lrucache.plugin_ctx(lrucache, ctx, nil, new_rr_obj, weighted_upstreams,
+                                                route_upstream_id)
     if not rr_up then
         core.log.error("lrucache roundrobin failed: ", err)
         return 500
```

Reply:
try
```lua   
 local rr_up, err = core.lrucache.plugin_ctx(lrucache, ctx, route_upstream_id, new_rr_obj, weighted_upstreams,
                                                route_upstream_id)
```

Reply:
update.
I know that using `core.lrucache.plugin_ctx` is more in line with the common way, but I don't understand the benefit of using `api_ctx.conf_type` and `api_ctx.conf_id` as lrucache keys over `weighted_upstreams`.

Reply:
The default key should ensure if the configuration of the route changes, it will change too. So we can ensure either route_upstream_id or weighted_upstreams changes, the version will change.

The corruption of the route configuration is another story (which should be fixed)...

Reply:
> So we can ensure either route_upstream_id or weighted_upstreams changes, the version will change.

got!